### PR TITLE
Fix multipart

### PR DESCRIPTION
**Pull request by smallhive**
Ref https://github.com/nspcc-dev/neofs-testcases/pull/906

### CODE REVIEW

File: api/layer/multipart_upload.go [hunk 1]
```diff
@@ -606,6 +612,7 @@ func (n *layer) reUploadPart(ctx context.Context, uploadParams UploadPartParams,
 	uploadParams.Size = int64(obj.PayloadSize())
 	uploadParams.Reader = bytes.NewReader(obj.Payload())
 
+	n.log.Error("reUploadPart", zap.String("oid", id.String()), zap.Uint64("payload size", obj.PayloadSize()))
```
Comment (roman-khimov) [inline api/layer/multipart_upload.go:668]: Debug.

File: internal/neofs/tree.go [hunk 2]
```diff
@@ -284,15 +284,17 @@ func newPartInfo(node NodeResponse) (*data.PartInfo, error) {
 		case homoHashKV:
 			partInfo.HomoHash = []byte(value)
 		case elementsKV:
-			elements := strings.Split(value, ",")
-			partInfo.Elements = make([]data.LinkObjectPayload, len(elements))
-			for i, e := range elements {
-				var element data.LinkObjectPayload
-				if err = element.Unmarshal(e); err != nil {
-					return nil, fmt.Errorf("invalid element: %w", err)
+			if value != "" {
```
Comment (roman-khimov) [inline internal/neofs/tree.go:287]: Seems unrelated.

File: api/layer/multipart_upload.go [hunk 3]
```diff
@@ -1287,26 +1325,13 @@ func (n *layer) uploadPartAsSlot(ctx context.Context, params uploadPartAsSlotPar
...
-	}
-
-	id, elements, err = n.manualSlice(ctx, params.bktInfo, prm, splitFirstID, splitPreviousID, *chunk, params.payloadReader)
-	if isReturnToPool {
-		n.buffers.Put(chunk)
+		Payload:      params.payloadReader,
+		PayloadSize:  uint64(params.decSize),
 	}
 
+	id, objHashBts, err := n.objectPutAndHash(ctx, prm, params.bktInfo)
```
Comment (roman-khimov) [inline api/layer/multipart_upload.go:1332]: So, no links to first ID?

File: api/layer/multipart_upload.go [hunk 4]
```diff
@@ -595,14 +595,67 @@ func (n *layer) reUploadFollowingParts(ctx context.Context, uploadParams UploadP
...
+func (n *layer) reUploadSegmentedPart(ctx context.Context, uploadParams UploadPartParams, part *data.PartInfo, bktInfo *data.BucketInfo, multipartInfo *data.MultipartInfo) error {
+	completePartPayload := make([]byte, 0, n.neoFS.MaxObjectSize())
+
+	for _, element := range part.Elements {
+		elementObj, err := n.objectGet(ctx, bktInfo, element.OID)
+		if err != nil {
+			return fmt.Errorf("get part oid=%s, emelent oid=%s: %w", part.OID.String(), element.OID.String(), err)
+		}
+
+		completePartPayload = append(completePartPayload, elementObj.Payload()...)
```
Comment (roman-khimov) [inline api/layer/multipart_upload.go:621]: 10G object?

Comment (smallhive) [reply api/layer/multipart_upload.go:621]: This is a huge problem because we have to know the whole part size before uploading. I am thinking about trying to add some extra fields in the meta, to get this info it

File: api/layer/multipart_upload.go (hunk 1 above)
Comment (smallhive) [reply api/layer/multipart_upload.go]: Updated

File: api/layer/multipart_upload.go (hunk 3 above)
Comment (smallhive) [reply api/layer/multipart_upload.go:1332]: No, we upload Slot just as a regular object

File: internal/neofs/tree.go (hunk 2 above)
Comment (smallhive) [reply internal/neofs/tree.go:287]: Panics otherwise, in case of the slot the part doesn't contain elements

File: api/layer/multipart_upload.go (hunk 4 above)
Comment (smallhive) [reply api/layer/multipart_upload.go]: UPD: Replaced with io.Pipe. It should be less memory aggressive

File: api/layer/multipart_upload.go (hunk 1 above)
Comment (roman-khimov) [reply api/layer/multipart_upload.go]: Expected to a part of the first commit.

File: api/layer/multipart_upload.go [hunk 5]
```diff
@@ -595,9 +596,83 @@ func (n *layer) reUploadFollowingParts(ctx context.Context, uploadParams UploadP
...
+
+func (n *layer) reUploadSegmentedPart(ctx context.Context, uploadParams UploadPartParams, part *data.PartInfo, bktInfo *data.BucketInfo, multipartInfo *data.MultipartInfo) error {
+	var (
+		eg                     errgroup.Group
+		pipeReader, pipeWriter = io.Pipe()
+	)
+
+	eg.Go(func() error {
+		for _, element := range part.Elements {
+			elementObj, err := n.objectGet(ctx, bktInfo, element.OID)
```
Comment (roman-khimov) [inline api/layer/multipart_upload.go:621]: I think it's overcomplicated. If you have a proper object, you may not care about its individual elements any more, you can just issue a GET for it, let node assemble individual parts, get an io.Reader and do io.Copy from it into larger object io.Writer.

Comment (smallhive) [reply api/layer/multipart_upload.go:621]: It is a key. If the part has elements, we slice it manually. This means we don't have a proper object and have to do all these actions

File: api/layer/multipart_upload.go (hunk 1 above)
Comment (smallhive) [reply api/layer/multipart_upload.go]: Fixed

File: api/layer/multipart_upload.go [hunk 6]
```diff
@@ -589,14 +596,100 @@ func (n *layer) reUploadFollowingParts(ctx context.Context, uploadParams UploadP
...
+	eg.Go(func() error {
+		var (
+			err        error
+			elementObj *object.Object
+		)
+
+		for _, element := range part.Elements {
+			elementObj, err = n.objectGet(ctx, bktInfo, element.OID)
+			if err != nil {
+				err = fmt.Errorf("get part oid=%s, emelent oid=%s: %w", part.OID.String(), element.OID.String(), err)
```
Comment (roman-khimov) [inline api/layer/multipart_upload.go:628]: emelent?

File: api/layer/multipart_upload.go [hunk 7]
```diff
@@ -589,14 +596,100 @@ func (n *layer) reUploadFollowingParts(ctx context.Context, uploadParams UploadP
...
+				// no return intentionally.
+			}
+		}
+
+		if err != nil {
+			pipeCloseErr := pipeWriter.Close()
+			return fmt.Errorf("pipe: %w. PipeClose: %w", err, pipeCloseErr)
+		}
+
+		if err = pipeWriter.Close(); err != nil {
```
Comment (roman-khimov) [inline api/layer/multipart_upload.go:662]: You can always just `pipeCloseErr := pipeWriter.Close()` and then deal with errors.

File: api/layer/multipart_upload.go (hunk 6 above)
Comment (smallhive) [reply api/layer/multipart_upload.go]: fixed

File: api/layer/multipart_upload.go (hunk 7 above)
Comment (smallhive) [reply api/layer/multipart_upload.go]: I hope I understood in the right way

File: api/layer/multipart_upload.go [hunk 8]
```diff
@@ -589,9 +596,90 @@ func (n *layer) reUploadFollowingParts(ctx context.Context, uploadParams UploadP
...
+					zap.String("uploadID", multipartInfo.UploadID),
+					zap.Int("partNumber", part.Number),
+					zap.String("part.OID", part.OID.String()),
+					zap.String("part element OID", element.OID.String()),
+				)
+				// no return intentionally.
+			}
+		}
+
+		if pipeCloseErr := pipeWriter.Close(); pipeCloseErr != nil {
```
Comment (roman-khimov) [inline api/layer/multipart_upload.go:657]: So if `pipeCloseErr == nil` and `err != nil` we're fine? I want to deduplicate `Close()`, but errors should be passed properly still.

Comment (smallhive) [reply api/layer/multipart_upload.go]: Updated

**Review by roman-khimov — APPROVED** [review]
